be/src/storage/lake/lake_primary_index.cpp

```diff
@@ -168,4 +174,58 @@ Status LakePrimaryIndex::_do_lake_load(TabletManager* tablet_mgr, const TabletMe
     return Status::OK();
 }
 
+Status LakePrimaryIndex::apply_opcompaction(const TabletMetadata& metadata,
+                                            const TxnLogPB_OpCompaction& op_compaction) {
+    if (!_enable_persistent_index || metadata.persistent_index_type() == PersistentIndexTypePB::LOCAL) {
+        return Status::OK();
+    }
+
+    auto* lake_persistent_index = dynamic_cast<LakePersistentIndex*>(_persistent_index.get());
+    if (lake_persistent_index != nullptr) {
+        return dynamic_cast<LakePersistentIndex*>(_persistent_index.get())->apply_opcompaction(op_compaction);
+    } else {
+        return Status::InternalError("Persistent index is not a LakePersistentIndex.");
+    }
+}
+
+Status LakePrimaryIndex::commit(const TabletMetadataPtr& metadata, MetaFileBuilder* builder) {
+    if (!_enable_persistent_index) {
+        return Status::OK();
+    }
+    switch (metadata->persistent_index_type()) {
+    case PersistentIndexTypePB::LOCAL: {
+        // only take affect in local persistent index
+        PersistentIndexMetaPB index_meta;
+        DataDir* data_dir = StorageEngine::instance()->get_persistent_index_store(_tablet_id);
+        RETURN_IF_ERROR(TabletMetaManager::get_persistent_index_meta(data_dir, _tablet_id, &index_meta));
+        RETURN_IF_ERROR(PrimaryIndex::commit(&index_meta));
+        RETURN_IF_ERROR(TabletMetaManager::write_persistent_index_meta(data_dir, _tablet_id, index_meta));
+        // Call `on_commited` here, which will remove old files is safe.
+        // Because if publish version fail after `on_commited`, index will be rebuild.
+        return on_commited();
+    }
+    case PersistentIndexTypePB::CLOUD_NATIVE: {
+        dynamic_cast<LakePersistentIndex*>(_persistent_index.get())->commit(builder);
```

Comment:
a lot of dynamic_cast, a smell of design problem here. PersistentIndex is trying do abstraction, but still need to know details about the concrete class info. must be something wrong.

Reply:
does it mean the cloudnative pk index commit always success?

Reply:
> does it mean the cloudnative pk index commit always success?

yes, cloudnative pk index commit only set the `PersistentIndexSstableMetaPB ` to `TabletMetadataPB`.

Reply:
Make sure _persistent_index isn't nullptr